### PR TITLE
Use bash strict mode in all CI scripts

### LINKED ISSUE #2010: Whitespace Check did not detect - no new line at EOF
### What happened?

Q. I see whilespace-check job is not erroring out for FATAL error 
Q. why "Not a valid object name origin/main" for PR https://github.com/openxla/stablehlo/pull/2007 ?  Dont see this an issue for other recent PRs

```
Run ./build_tools/github_actions/lint_whitespace_checks.sh
fatal: Not a valid object name origin/main
Checking whitespace:
fatal: ambiguous argument '': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
  
fatal: ambiguous argument '': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
tail: cannot open 'bash' for reading: No such file or directory
fatal: ambiguous argument '': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
No whitespace issues found.

```

### Version information

_No response_

### PR DESCRIPTION
**Pull request by mlevesquedion**
"Bash strict mode" is a colloquial expression for a set of settings that
makes Bash stricter (safer). See e.g. https://gist.github.com/robin-a-meade/58d60124b88b60816e8349d1e3938615

TLDR: to avoid errors from being silently ignored, we should set errexit
(exit script on error), nounset (exit script when an undefined variable
is encountered) and pipefail (propagate error codes in pipelines) in all
our scripts.

Unfortunately, `errexit` is in fact not enough to exit when any error
occurs, some errors are swallowed: https://dev.to/banks/stop-ignoring-errors-in-bash-3co5

Because of this, I changed the whitespace script to gather the files in
a different way (I copied the approach from the clang-format script), such
that the script will exit when an error occurs, instead of continuing to run.

The whitespace script was missing a fetch which was causing `git
merge-base` to fail. Maybe our other scripts should also be updated to
only run on changed files.

Interestingly, GitHub already sets `-e` (`errexit`) when running
scripts, but it's best to enforce it in the script itself:

```
Run ./build_tools/github_actions/lint_whitespace_checks.sh
  ./build_tools/github_actions/lint_whitespace_checks.sh
  shell: /usr/bin/bash -e {0}
```

Fixes #2010

### CODE REVIEW

**Comment by mlevesquedion** [comment]
Converting to draft until I get a some time to fix the failures.

**Review by abhigunj — APPROVED** [review]
